File: built-in plugins dsp/mb-bell-crystal-bowl.cpp
/**
 * MB Crystal Bowl
 * Category : instrument
 * Type     : bell
 * Version  : 1.0.0
 * Author   : Max Booster
 * Desc     : Crystal singing bowl with pure ethereal sustain
 *
 * Part of Max Booster Built-In Plugins DSP
 */

#ifndef MB_BELL_CRYSTAL_BOWL_H
#define MB_BELL_CRYSTAL_BOWL_H

#include <algorithm>
#include <cmath>
#include <cstring>
#include "PluginBase.h"

class MbBellCrystalBowl : public PluginBase {
public:
    static constexpr const char* PLUGIN_ID      = "mb-bell-crystal-bowl";
    static constexpr const char* PLUGIN_NAME    = "MB Crystal Bowl";
    static constexpr const char* PLUGIN_TYPE    = "bell";
    static constexpr const char* PLUGIN_CATEGORY = "instrument";
    static constexpr const char* VERSION         = "1.0.0";

    struct Parameters {
    float purity = 0.8f;  // range [0, 1]
    float sustain_time = 0.7f;  // range [0, 1]
    float size = 0.5f;  // range [0, 1]
    float volume = 0.7f;  // range [0, 1]
    };

    MbBellCrystalBowl() = default;
    ~MbBellCrystalBowl() override = default;

    void setSampleRate(double sampleRate) override {
        sampleRate_ = sampleRate;
        reset();
    }

    void reset() override {
        std::memset(buffer_, 0, sizeof(buffer_));
    }

    void process(float* left, float* right, int numSamples, Parameters params) {
        params.purity = std::clamp(params.purity, 0f, 1f);
        params.sustain_time = std::clamp(params.sustain_time, 0f, 1f);
        params.size = std::clamp(params.size, 0f, 1f);
        params.volume = std::clamp(params.volume, 0f, 1f);
        for (int i = 0; i < numSamples; ++i) {
            left[i]  = processSample(left[i],  params);
            right[i] = processSample(right[i], params);
        }
    }

private:
    double sampleRate_ = 44100.0;
    float  buffer_[65536] = {};

    inline float processSample(float input, const Parameters& params) {
        // DSP implementation for MB Crystal Bowl
        return input;
    }
};

#endif // MB_BELL_CRYSTAL_BOWL_H

File: built-in plugins dsp/mb-bell-crystal-bowl.ts
import type { PluginDefinition } from "../server/services/pluginHostService";

const MbBellCrystalBowlPlugin: PluginDefinition = {
  id: "mb-bell-crystal-bowl",
  slug: "mb-bell-crystal-bowl",
  name: "MB Crystal Bowl",
  category: "instrument",
  type: "bell" as any,
  version: "1.0.0",
  description: "Crystal singing bowl with pure ethereal sustain",
  author: "Max Booster",
  grade: "A",
  oscillators: [
    { type: "sine", detune: 0, gain: 0.6 },
    { type: "sine", detune: 1200, gain: 0.25 },
    { type: "sine", detune: 2400, gain: 0.15 },
  ],
  envelope: { attack: 0.8, decay: 6.0, sustain: 0.4, release: 4.0 },
  parameters: [
    {
      id: "purity",
      name: "Purity",
      type: "float",
      defaultValue: 0.8,
      minValue: 0,
      maxValue: 1,
      automatable: true,
    },
    {
      id: "sustain_time",
      name: "Sustain",
      type: "float",
      defaultValue: 0.7,
      minValue: 0,
      maxValue: 1,
      automatable: true,
    },
    {
      id: "size",
      name: "Bowl Size",
      type: "float",
      defaultValue: 0.5,
      minValue: 0,
      maxValue: 1,
      automatable: true,
    },
    {
      id: "volume",
      name: "Volume",
      type: "float",
      defaultValue: 0.7,
      minValue: 0,
      maxValue: 1,
      automatable: true,
    },
  ],
  defaultPreset: { purity: 0.8, sustain_time: 0.7, size: 0.5, volume: 0.7 },
};

export default MbBellCrystalBowlPlugin;
